src/steamship/agents/service/agent_service.py

```diff
@@ -146,6 +148,20 @@ def run_agent(self, agent: Agent, context: AgentContext):
 
         while not isinstance(action, FinishAction):
             self.run_action(agent=agent, action=action, context=context)
+
+            # If this action wishes to end the loop, comply with that wish
+            if action.return_direct:
```

Comment:
aside (take it or leave it): my api spidey-sense is tingling seeing this code block. i think seeing this, i'd prefer something like `action.is_final` (or even `if isinstance(action, DirectReturnAction)`) for clarity (as `Action`s aren't really "returning" anything per se).